### ISSUE TITLE
[OSDI'20] KungFu: Making Training in Distributed Machine Learning Adaptive

### ISSUE THREAD

**Issue by ganler** [ganler/ResearchReading#27]
Milo's latest work on adaptive training.

> KungFu, a distributed ML training library that is designed to adapt configuration parameters at runtime.

**Comment by ganler** [comment]
## Motivation

What we do currently:

- Set parameters(system parameters and training parameters at deploy time).

Question: Can we make it adaptive? e.g., can we have different batch sizes or learning rates during runtime? And can we adaptively change our system resources during training? 

Kungfu is a system that allows users to define their adaptive training strategy via high-level *Adaptive Policies*(AP).

**Comment by ganler** [comment]
## Key Contrib.

1. The expressiveness of AP.
2. Efficient monitoring.
3. Configurations are of distributed mechanisms.

**Comment by ganler** [comment]
There are some ML papers from OpenAI, NVIDIA, etc. arguing that to better train our models, we need to change the batch size adaptively based on the monitored feature(e.g., gradient noise scale). And that's why KungFu also aims at speeding up the monitoring performance.

Also, we can change the system configurations(i.g., network topology, computation scheduling) given the monitored worker utilization or work communication rate. However, in distributed training, GBs of gradients are produced in 1 second, shipping such amount of data to the monitoring system like TensorBoard, MLFlow, or Prometheus consumes substantial network bandwidth consumption. (or you may agree with BytePS which regards CPU servers free that the extreme bandwidth consumption of metrics server is inconsequential. lol)

Adaptation policies make it possible to train models faster, more accurately, in a "fluent" manner(that's why this paper also addresses the expressiveness of policies).

**Comment by ganler** [comment]
## AP interface

> Policy can be used as callbacks.

User-def functions:

- `before_train` + `after_train` +`after_step`
- `before_epoch` + `after_epoch`

There's also a line of built-in functions in KungFu about **communication**, **monitoring**, and **adaptation**.

**Comment by ganler** [comment]
## Collective communication

In a decentralized fashion.

> While Horovod employs a centralized fashion. (Maybe they don't have requirements for monitoring big chunk of gradients)

### Named Collective Message

Entry: [Key] [Data]

### Named Collective States

Entry: [Control] [Data]

## Adapting parameters of workers

> This means to change configuration parameters at runtime.

Existing frameworks for adaptive training:
- Horovod Elastic: 
- Auto-Scaling: 

- require users to use a dynamic execution mode of the ML framework.
- Only support limited parameters, e.g., the number of workers.

"The core idea is that, instead of providing configuration parameters as static parameters when compiling the dataflow, KungFu adds parameters as computational configuration operators to the dataflow graph."

This idea makes monitoring (i) efficient as the dimension of gradients is reduced once gradients are produced by the intermediate layer, and (ii) adaptive as the configuration operators can alter the outputs dynamically by reading the configuration parameters.
